src/providers/haskell.rs

```diff
@@ -0,0 +1,76 @@
+use std::collections::BTreeMap;
+
+use serde::Deserialize;
+
+use crate::nixpacks::{
+    nix::Pkg,
+    phase::{BuildPhase, InstallPhase, SetupPhase, StartPhase},
+};
+
+use super::Provider;
+
+pub struct HaskellStackProvider {}
+
+impl Provider for HaskellStackProvider {
+    fn name(&self) -> &str {
+        "haskell_stack"
+    }
+
+    fn detect(
+        &self,
+        app: &crate::nixpacks::app::App,
+        _env: &crate::nixpacks::environment::Environment,
+    ) -> anyhow::Result<bool> {
+        Ok(app.includes_file("package.yaml"))
+    }
+
+    fn setup(
+        &self,
+        _app: &crate::nixpacks::app::App,
+        _env: &crate::nixpacks::environment::Environment,
+    ) -> anyhow::Result<Option<crate::nixpacks::phase::SetupPhase>> {
+        Ok(Some(SetupPhase::new(vec![
+            Pkg::new("stack"),
+            Pkg::new("apt"),
+        ])))
+    }
+
+    fn install(
+        &self,
+        _app: &crate::nixpacks::app::App,
+        _env: &crate::nixpacks::environment::Environment,
+    ) -> anyhow::Result<Option<crate::nixpacks::phase::InstallPhase>> {
+        Ok(Some(InstallPhase::new("sudo apt-get update && sudo apt-get install -y libgmp-dev gcc binutils make && stack setup".to_string())))
```

Comment:
Can these packages be installed with Nix instead? `gcc` and `make` should be able to. What are `libgmp-dev` and `binutils` for?

Reply:
Nix doesn’t package any development libraries/headers. `libgmp-dev` is a floating-point math library needed by the Haskell compiler. `binutils` has things like `ar` that are needed to build libraries. `make` could probably be installed through Nix, but `gcc` needs to be installed using the same package manager as any libraries you’re trying to use, so that it can find them. Since `apt` is the only way to get development libraries, `gcc` needs to be installed through it too.

Reply:
Ah I see. I thinkkk this is okay. I'm hesitant to start having providers download external dependencies outside of Nix since we lose the ability to pin to a specific version of the Nix store. I think this is fine for now but we should be cautious about encouraging this moving forward.

If more than one provider can be matched at a time (future plans) then it would be nicer if the Haskell provider could depend on the AptProvider and specify the packages that way. Then at least the behaviour is consistent and more controller than just installing w/e packages in the install phase.